src/ReverseTree.ts

```diff
@@ -71,58 +66,98 @@ export class ReverseTree {
         after: parseItemID(checkpoint, afters[i]),
         id: parseItemID(checkpoint, ids[i]),
         value: values[i],
+        versionstamp: checkpoint.vs,
+        local: false,
       };
-      this.nodes[node.id] = node;
-      this.log.push(node);
+      this.nodes.set(node.id, node);
     }
-
-    this.count = this.log.length;
   }
 
   get(itemID: string): NodeValue | undefined {
-    if (this.nodes[itemID]) {
-      return this.nodes[itemID].value;
-    }
-    return undefined;
+    return this.nodes.get(itemID)?.value;
   }
 
-  insert(
-    after: 'root' | string,
-    value: NodeValue,
-    externalNewID?: string
-  ): string {
-    invariant(this.log);
+  insert(params: {
+    after: 'root' | string;
+    value: NodeValue;
+    externalNewID?: string;
+    localOrAck: LocalOrAck;
+  }): string {
+    const { after, value, externalNewID, localOrAck } = params;
     let id = externalNewID;
     if (!id) {
-      id = `${this.count}:${this.actor}`;
+      //  TODO: with list GC implemented, count should be incremented higher than
+      //  any node id instead of based on current number of active nodes
+      id = `${this.nodes.size}:${this.session}`;
+    }
+    if (this.nodes.has(id)) {
+      if (this.nodes.get(id)?.after === undefined) {
+        this.nodes.get(id)!.after = after;
+      }
+      this.put(id, value, localOrAck);
+      return id;
     }
-    this.count++;
 
     const node: Node = {
       after,
       value,
       id,
+      local: localOrAck === 'local',
+      versionstamp:
+        (typeof localOrAck === 'object' && localOrAck.versionstamp) ||
+        undefined,
     };
-    this.nodes[id] = node;
-    this.log.push(node);
+    this.nodes.set(id, node);
     return id;
   }
 
-  put(itemID: string, value: NodeValue) {
-    if (!!this.nodes[itemID]) {
-      this.nodes[itemID].value = value;
+  put(itemID: string, value: NodeValue, localOrAck: LocalOrAck) {
+    const existing = this.nodes.get(itemID);
+    if (!existing) {
+      //  handle out of order lput/ldel and lins
+      this.nodes.set(itemID, {
+        after: undefined,
+        value,
+        versionstamp:
+          (typeof localOrAck === 'object' && localOrAck.versionstamp) ||
+          undefined,
+        id: itemID,
+        local: localOrAck === 'local',
+      });
+      return;
+    }
+
+    if (localOrAck === 'local') {
+      existing.value = value;
+      existing.local = true;
+      return;
+    }
+
+    const versionstamp = localOrAck.versionstamp;
+    if (
+      versionstamp &&
+      existing.versionstamp &&
+      isOlderVS(versionstamp, existing.versionstamp)
+    ) {
+      return;
     }
+
+    const isLocalAck = existing.local && localOrAck.ack;
+    if (isLocalAck) {
+      //  ignore value
+    } else {
+      existing.value = value;
+      existing.local = false;
+    }
```

Comment:
Out of curiosity, is this format to make it more obvious we're ignoring? Why not just do `if (!isLocalAck)`? 

Not a request, it's cool, just curious

Reply:
That's correct, I thought it looked better than

```
    //  ignore value if isLocalAck
    if (!isLocalAck) {
      existing.value = value;
      existing.local = false;
    }
```